File: src/test/resources/py/rampup.py
from __future__ import print_function, absolute_import, with_statement
#
from loadbuilder import init as loadBuilderInit
from chain import build as chainBuild
from chain import execute as chainExecute
#
from java.lang import InterruptedException, Long, Short, Thread, Throwable, NumberFormatException
#
from org.apache.logging.log4j import Level, LogManager
#
from com.emc.mongoose.core.api.persist import Markers
from com.emc.mongoose.core.impl.persist import TraceLogger
from com.emc.mongoose.core.impl.util import ThreadContextMap
from com.emc.mongoose.core.impl.util import RunTimeConfig
#
LOG = LogManager.getLogger()
#
def init():
	runTimeConfig = RunTimeConfig.getContext()
	LOG.debug(Markers.MSG, "Setting the metric update period to zero for chain scenario")
	runTimeConfig.set("run.metrics.period.sec", 0)
	#
	loadTypesChain = runTimeConfig.getStringArray("scenario.chain.load")
	LOG.info(Markers.MSG, "Load types chain: {}", loadTypesChain)
	listSizes = runTimeConfig.getStringArray("scenario.rampup.sizes")
	LOG.info(Markers.MSG, "Data sizes: {}", listSizes)
	listThreadCounts = runTimeConfig.getStringArray("scenario.rampup.thread.counts")
	LOG.info(Markers.MSG, "Thread counts: {}", listThreadCounts)
	return loadTypesChain, listSizes, listThreadCounts
#
def execute(loadBuilder, rampupParams=((),(),())):
	loadTypesChain = rampupParams[0]
	listSizes = rampupParams[1]
	listThreadCounts = rampupParams[2]
	for index, dataItemSizeStr in enumerate(listSizes):
		try:
			dataItemSize = Long(RunTimeConfig.toSize(dataItemSizeStr))
			for threadCountStr in listThreadCounts:
				nextChain = None
				try:
					threadCount = Short.valueOf(threadCountStr)
				except NumberFormatException as e:
					TraceLogger.failure(Markers.ERR, Level.WARN, e, "Failed to parse the next thread count")
				try:
					LOG.info(Markers.PERF_SUM, "---- Step {}x{} start ----", threadCount, dataItemSizeStr)
					ThreadContextMap.putValue("currentSize", dataItemSizeStr + "-" + str(index))
					ThreadContextMap.putValue("currentThreadCount", str(threadCount))
					nextChain = chainBuild(
						loadBuilder, loadTypesChain, False, True, dataItemSize, dataItemSize, threadCount
					)
					chainExecute(nextChain, False)
					LOG.debug(Markers.MSG, "---- Step {}x{} finish ----", threadCount, dataItemSizeStr)
				except InterruptedException as e:
					raise e
				except Throwable as e:
					TraceLogger.failure(LOG, Level.ERROR, e, "Chain execution failure")
		except InterruptedException:
			break
		except Throwable as e:
			TraceLogger.failure(LOG, Level.ERROR, e, "Determining the next data item size failure")
#
if __name__ == "__builtin__":
	execute(loadBuilder=loadBuilderInit(), rampupParams=init())
	LOG.info(Markers.MSG, "Scenario end")

File: src/main/java/com/emc/mongoose/run/examples/Rampup.java
package com.emc.mongoose.run.examples;
//
import com.emc.mongoose.common.conf.Constants;
import com.emc.mongoose.common.conf.RunTimeConfig;
import com.emc.mongoose.common.conf.SizeUtil;
import com.emc.mongoose.common.log.LogUtil;
import com.emc.mongoose.common.log.Markers;
//
import com.emc.mongoose.core.api.load.builder.LoadBuilder;
//
import com.emc.mongoose.run.cli.HumanFriendly;
import com.emc.mongoose.run.examples.shared.LoadBuilderFactory;
//
import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.ThreadContext;
//
import java.nio.file.Paths;
import java.rmi.RemoteException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;
/**
 Created by kurila on 11.06.15.
 */
public class Rampup
implements Runnable {
	//
	static {
		try {
			LogUtil.init();
			RunTimeConfig.initContext();
		} catch(final Exception e) {
			e.printStackTrace(System.err);
		}
	}
	private final static Logger LOG = LogManager.getLogger();
	//
	private final Map<String, Chain> loadStepsMap = new HashMap();
	//
	public Rampup(
		final LoadBuilder loadBuilder, final long timeOut, final TimeUnit timeUnit,
		final String loadTypeSeq[], final String sizeSeq[], final String threadCountSeq[]
	) {
		String nextSizeStr, nextStepName;
		long nextSize;
		for(int i = 0; i < sizeSeq.length; i++) {
			nextSizeStr = sizeSeq[i];
			for(final String nextThreadCountStr : threadCountSeq) {
				ThreadContext.put("currentSize", nextSizeStr + "-" + i);
				ThreadContext.put("currentThreadCount", nextThreadCountStr);
				nextSize = SizeUtil.toSize(nextSizeStr);
				nextStepName = nextThreadCountStr + "x" + nextSizeStr;
				LOG.debug(Markers.MSG, "Build the next step load chain: \"{}\"", nextStepName);
				try {
					loadBuilder
						.setMinObjSize(nextSize)
						.setMaxObjSize(nextSize)
						.setThreadsPerNodeDefault(Short.parseShort(nextThreadCountStr));
					loadStepsMap.put(
						nextStepName, new Chain(loadBuilder, timeOut, timeUnit, loadTypeSeq, false)
					);
				} catch(final RemoteException e) {
					LogUtil.exception(LOG, Level.WARN, e, "Failed to apply rampup params remotely");
				}
			}
		}
	}
	//
	@Override
	public final void run() {
		for(final String nextStepName : loadStepsMap.keySet()) {
			LOG.info(Markers.PERF_SUM, "---- Step {} start ----", nextStepName);
			loadStepsMap.get(nextStepName).run();
		}
	}
	//
	public static void main(final String... args) {
		try {
			//
			final RunTimeConfig runTimeConfig = RunTimeConfig.getContext();
			runTimeConfig.loadPropsFromJsonCfgFile(
				Paths.get(RunTimeConfig.DIR_ROOT, Constants.DIR_CONF)
					.resolve(RunTimeConfig.FNAME_CONF)
			);
			runTimeConfig.loadSysProps();
			// load the config from CLI arguments
			final Map<String, String> properties = HumanFriendly.parseCli(args);
			if(!properties.isEmpty()) {
				LOG.debug(Markers.MSG, "Overriding properties {}", properties);
				RunTimeConfig.getContext().overrideSystemProperties(properties);
			}
			//
			LOG.info(Markers.MSG, RunTimeConfig.getContext().toString());
			//
			final LoadBuilder loadBuilder = LoadBuilderFactory.getInstance();
			final long timeOut = runTimeConfig.getLoadLimitTimeValue();
			final TimeUnit timeUnit = runTimeConfig.getLoadLimitTimeUnit();
			//
			final String[] loadTypeSeq = runTimeConfig.getScenarioChainLoad();
			final String[] sizeSeq = runTimeConfig.getScenarioRampupSizes();
			final String[] threadCountSeq = runTimeConfig.getScenarioRampupThreadCounts();
			final Rampup rampupScenario = new Rampup(
				loadBuilder, timeOut, timeUnit, loadTypeSeq, sizeSeq, threadCountSeq
			);
			//
			rampupScenario.run();
		} catch(final Exception e) {
			e.printStackTrace(System.err);
			LogUtil.exception(LOG, Level.ERROR, e, "Scenario failed");
		}
	}
}
